File: APDs/wiringpi/APD.cpp
#include "2591a.h" //header including memory information

#include <wiringPi.h>
#include <stdio.h>
#include <softTone.h>    //piezo lib.
#include <stdlib.h>      //used in handling sensor data
#include <stdint.h>      //used in handling sensor data
#include <wiringPiI2C.h> //Wiringpi i2c lib
#include <unistd.h>      //used sensor data storage

//pin number definition
#define PIEZO 1

#define LBTN 4
#define RBTN 5
#define ABTN 2

#define RLED 25
#define GLED 24
#define BLED 22

#define DHT_PIN 0

#define MAX_TIMINGS 200 //temp sensor timing

int data[5] = {0, 0, 0, 0, 0}; //temp data array

void read_dht_data()
{
    uint8_t laststate = HIGH;
    uint8_t counter = 0;
    uint8_t j = 0, i;

    data[0] = data[1] = data[2] = data[3] = data[4] = 0;

    /* pull pin down for 18 milliseconds */
    pinMode(DHT_PIN, OUTPUT);
    digitalWrite(DHT_PIN, LOW);
    delay(10);

    /* prepare to read the pin */
    pinMode(DHT_PIN, INPUT);

    /* detect change and read data */
    for (i = 0; i < MAX_TIMINGS; i++)
    {
	counter = 0;
	while (digitalRead(DHT_PIN) == laststate)
	{
	    counter++;
	    delayMicroseconds(1);
	    if (counter == 255)
	    {
		break;
	    }
	}
	laststate = digitalRead(DHT_PIN);

	if (counter == 255)
	    break;

	/* ignore first 3 transitions */
	if ((i >= 4) && (i % 2 == 0))
	{
	    /* shove each bit into the storage bytes */
	    data[j / 8] <<= 1;
	    if (counter > 16)
		data[j / 8] |= 1;
	    j++;
	}
    }

    /*
	 * check we read 40 bits (8bit x 5 ) + verify checksum in the last byte
	 * print it out if data is good
	 */
    if ((j >= 40) &&
	(data[4] == ((data[0] + data[1] + data[2] + data[3]) & 0xFF)))
    {
	float h = (float)((data[0] << 8) + data[1]) / 10;
	if (h > 100)
	{
	    h = data[0]; // for DHT11
	}
	float c = (float)(((data[2] & 0x7F) << 8) + data[3]) / 10;
	if (c > 125)
	{
	    c = data[2]; // for DHT11
	}
	if (data[2] & 0x80)
	{
	    c = -c;
	}
	float f = c * 1.8f + 32;
	printf("Humidity = %.1f %% Temperature = %.1f *C (%.1f *F)\n", h, c, f);
    }
}

#define TSL2561_ADDR_LOW (0x29)
#define TSL2561_ADDR_FLOAT (0x39)
#define TSL2561_ADDR_HIGH (0x49)

void getLux(int fd)
{
    Adafruit_TSL2591 tsl = Adafruit_TSL2591(2591); // pass in a number for the sensor identifier (for your use later)

    uint16_t ch0, ch1;
    uint32_t visible_and_ir, lux;

    visible_and_ir = tsl.Adafruit_TSL2591::getFullLuminosity(fd);

    visible_and_ir = tsl.Adafruit_TSL2591::getFullLuminosity(fd);

    // Reads two byte value from channel 0 (visible + infrared)
    ch0 = (visible_and_ir & 0xFFFF);
    // Reads two byte value from channel 1 (infrared)
    ch1 = (visible_and_ir >> 16);

    lux = tsl.Adafruit_TSL2591::calculateLux(ch0, ch1);

    //    printf("TSL2591 Full Spectrum: %zu\n",ch0);
    //    printf("TSL2591 Infrared: %zu\n",ch1);
    //   if ( ch0 >= ch1 )  printf("TSL2591 Visible: %zu\n",ch0-ch1);

    if (lux < 20)
    {
	digitalWrite(BLED, 0);
    }
    else
    {
	digitalWrite(BLED, 1);
    }
} //output : LED  // normal = 40 ~ 200

int change_mode(int status)
{ // toggle status
    if (status == 1)
    {
	digitalWrite(GLED, 1);
	printf("de-activate\n");
	return 0;
    }
    if (status == 0)
    {
	digitalWrite(GLED, 0);
	printf("activate\n");
	return 1;
    }
    printf("!!!!!!!!!!!!!error!!!!!!!!!\n\n");
    return 0;
}

void pushbutton_left()
{ // left button : beep 1sec
    softToneCreate(PIEZO);
    softToneWrite(PIEZO, 400);
    delay(400);
    softToneWrite(PIEZO, 0);
}

void pushbutton_right()
{ // right button : beep 2sec
    softToneCreate(PIEZO);
    softToneWrite(PIEZO, 400);
    delay(400);
    softToneWrite(PIEZO, 800);
    delay(400);
    softToneWrite(PIEZO, 0);
}

int main()
{
    //Initializing
    wiringPiSetup();
    pinMode(PIEZO, OUTPUT);
    pinMode(DHT_PIN, INPUT);

    pinMode(LBTN, INPUT);
    pinMode(RBTN, INPUT);
    pinMode(ABTN, INPUT);

    pinMode(RLED, OUTPUT);
    pinMode(GLED, OUTPUT);
    pinMode(BLED, OUTPUT);

    digitalWrite(RLED, 1);
    digitalWrite(GLED, 1);
    digitalWrite(BLED, 1);

    int fd = wiringPiI2CSetup(0x29);

    printf("Initialized!\n\n");

    int count = 0;
    int status = 0;

    while (1)	// loop
    {
		//button handling
	if (digitalRead(ABTN) == LOW)
	{
	    status = change_mode(status);
	    delay(200);
	}

	if (status == 1)
	{
	    if (digitalRead(LBTN) == LOW)
	    {
		pushbutton_left();
	    }
	    if (digitalRead(RBTN) == LOW)
	    {
		pushbutton_right();
	    }
	}

	count++;

	if (count > 15)	// light sensor output : LED
	{
	    read_dht_data();
	    count = 0;
	}
	getLux(fd);
    }
    return 0;
}

File: APDs/node/APD.js
var five = require("johnny-five");
var Raspi = require("raspi-io");
var temp = require("node-dht-sensor");
var tsl2591 = require("tsl2591");
var Gpio = require('onoff').Gpio;
var sleep = require('sleep');

var board = new five.Board({    //Initialize with J5-Rpio interface
    io: new Raspi()
});

board.on("ready", function () {
    var status = 1; // turn to 0 when initialized

    //pins Initialize
    var RLED = new five.Led("P1-37");
    var GLED = new five.Led("P1-35");

    var ABTN = new five.Button("P1-13");
    var LBTN = new five.Button("P1-16");
    var RBTN = new five.Button("P1-18");

    var piezo = new Gpio(18, 'out');
    var BLED = new Gpio(6, 'out');
    var light = new tsl2591({ device: "/dev/i2c-1" }); //i2c Initialize


    //on/off reversed
    RLED.on();
    GLED.on();
    BLED.writeSync(1);

    //toggle status switch
    ABTN.on("down", function () { //active button pushed

        if (status == 0) {
            status = 1;
            console.log("activated", status);
            GLED.off(); //on/off reversed
        }
        else if (status == 1) {
            status = 0;
            console.log("de-activated", status);
            GLED.on();  //on/off reversed
        }
    });

    LBTN.on("down", function () { //left button : beep 1sec
        console.log("LBTN");
        if (status == 1) {
            for (var i = 0; i < 300; i++) {
                piezo.writeSync(1);
                sleep.usleep(1000);
                piezo.writeSync(0);
                sleep.usleep(1000);
            };
        }
    });

    RBTN.on("down", function () { // right button : beep 2 sec
        console.log("RBTN");
        if (status == 1) {
            for (var i = 0; i < 300; i++) {
                piezo.writeSync(1);
                sleep.usleep(1000);
                piezo.writeSync(0);
                sleep.usleep(1000);
            };
            for (var i = 0; i < 300; i++) {
                piezo.writeSync(1);
                sleep.usleep(700);
                piezo.writeSync(0);
                sleep.usleep(700);
            };
        }
    });

    showLum();  //light sensor output(LED)
    showTmp();  //temperature sensor output(console)
});

light.init({ AGAIN: 0, ATIME: 1 }, function (err) { //light sensor Initialize
    if (err) {
        console.log(err);
        process.exit(-2);
    } else {
        console.log("Device ready");
    }
});

function showLum() { //light sensor output(LED)
    light.readLuminosity(function (err, data) {
        if (err) { console.log(err); }
        else {
            if (data.ir < 3) { //handle LED with onoff module, normal : 10 ~ 20
                BLED.writeSync(0);
            }
            else {
                BLED.writeSync(1);
            };
        }
    });
    setTimeout(showLum, 100); // delay 0.1sec
}

function showTmp() { //temp sensor output(console)
    temp.read(22, 17, function (err, temperature, humidity) {
        if (!err) {
            console.log("temp :" + temperature.toFixed(1) + "C,  " + "humidity:  " + humidity.toFixed(1) + "%");
        }  //directly read
    });
    setTimeout(showTmp, 5000); // delay 5sec
}
